db: inconsistent batch count panic

Seen while running the `tpcc/w=100/nodes=3/chaos=true` roachtest. From the stack, the panic is coming out of a batch application that occurred while replaying the WAL. The panic is indicating that the batch is supposed to have 7948101 entries, but applying the batch only saw 7948060 (a difference of 41 entries). Unfortunately, the underlying data for this cluster has been lost. Stressing this roachtest might reproduce.

Also, what in that roachtest is creating a batch with nearly 8m entries? That's shockingly high.

```
panic: pebble: inconsistent batch count: 7948060 vs 7948101 [recovered]
        panic: pebble: inconsistent batch count: 7948060 vs 7948101

goroutine 37 [running]:
panic(0x3d820a0, 0xc0009f7250)
        /usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc000b09358 sp=0xc000b092c8 pc=0x792752
github.com/cockroachdb/pebble.Open.func1(0xc000b09db8, 0xc001028f00, 0xc0001d2d00)
        /go/src/github.com/cockroachdb/pebble/open.go:76 +0x186 fp=0xc000b093b0 sp=0xc000b09358 pc=0x2bca336
runtime.call32(0x0, 0x4476758, 0xc000b09c08, 0x1800000018)
        /usr/local/go/src/runtime/asm_amd64.s:539 +0x3b fp=0xc000b093e0 sp=0xc000b093b0 pc=0x7c10ab
panic(0x3d820a0, 0xc0009f7250)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2 fp=0xc000b09470 sp=0xc000b093e0 pc=0x792642
github.com/cockroachdb/pebble.(*memTable).apply(0xc00027ed20, 0xc000b09988, 0x79470f, 0x0, 0x0)
        /go/src/github.com/cockroachdb/pebble/mem_table.go:215 +0x46c fp=0xc000b097c0 sp=0xc000b09470 pc=0x2bb08fc
github.com/cockroachdb/pebble.(*DB).replayWAL(0xc0001d2d00, 0x1, 0xc000b09d40, 0x4bdb7c0, 0x74b49a0, 0xc000746c80, 0x1f, 0x440, 0x794745, 0x0, ...)
        /go/src/github.com/cockroachdb/pebble/open.go:448 +0xba3 fp=0xc000b09a70 sp=0xc000b097c0 pc=0x2bbabf3
github.com/cockroachdb/pebble.Open(0x7ffe8e6fec29, 0x14, 0xc001028f00, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/pebble/open.go:231 +0xfa8 fp=0xc000b09da0 sp=0xc000b09a70 pc=0x2bb89b8
github.com/cockroachdb/cockroach/pkg/storage/engine.NewPebble(0x4b604a0, 0xc00060fd40, 0x0, 0x0, 0x0, 0x7ffe8e6fec29, 0x14, 0x0, 0x0, 0xc0001e6a80, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/pebble.go:405 +0x4cf fp=0xc000b09ef0 sp=0xc000b09da0 pc=0x2bf5dff
github.com/cockroachdb/cockroach/pkg/server.(*Config).CreateEngines(0xc0003d2808, 0x4b604a0, 0xc00060fd40, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/config.go:520 +0x1819 fp=0xc000b0a760 sp=0xc000b09ef0 pc=0x3060029
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start(0xc0003d2800, 0x4b604a0, 0xc000a8c870, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1333 +0xb91 fp=0xc000b0b5b8 sp=0xc000b0a760 pc=0x3070b71
github.com/cockroachdb/cockroach/pkg/cli.runStart.func3.2(0xc000c761b0, 0xc000010138, 0xc0002b24f0, 0x4b604a0, 0xc000a8c870, 0x0, 0x8b2eb5b, 0xed5dcdfec, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:708 +0x107 fp=0xc000b0be50 sp=0xc000b0b5b8 pc=0x31d95e7
github.com/cockroachdb/cockroach/pkg/cli.runStart.func3(0xc000010138, 0x4b604a0, 0xc000a8c870, 0x4bc9ac0, 0xc0002bd140, 0xc000c761b0, 0xc0002b24f0, 0x0, 0x8b2eb5b, 0xed5dcdfec, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:824 +0x181 fp=0xc000b0bf80 sp=0xc000b0be50 pc=0x31db2e1
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc000b0bf88 sp=0xc000b0bf80 pc=0x7c2dc1
created by github.com/cockroachdb/cockroach/pkg/cli.runStart
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:664 +0x9dd
```

That panic message is confusing. The numbers that are printed are not the batch count, but the expected vs actual seqnum after applying the batch. Another repro shows:

```
panic: pebble: inconsistent batch count: 5234363 vs 5234366
```

This corresponds to the following entry in a WAL which happens to be the last entry in the WAL:

```
7163714(12467) seq=5234361 count=5
    SET(0169f6b6757266746c00000000000003fa00,)
    SET(0169f6b6757266746c00000000000003fb00,)
```

So the count says `5`, but the batch seems to only contain 2 entries. Need to figure out where this discrepancy is coming from.

Another occurrence shows something similar:

```
panic: pebble: inconsistent batch count: 6401631 vs 640164
```

And the last entry in the WAL being replayed is:

```
13301746(2051) seq=6401625 count=21
    SET(c589c789f70b548800,)
    SET(c589c789f70b54880015f48500e11bdf6609,)
    SET(c589c78af705378800,)
    SET(c589c78af70537880015f48500e11bdf6609,)
    SET(c589c78bf703098800,)
    SET(c589c78bf70309880015f48500e11bdf6609,)
```

--- a/internal/record/record.go
+++ b/internal/record/record.go
@@ -165,8 +165,6 @@ type Reader struct {
 	// n is the number of bytes of buf that are valid. Once reading has started,
 	// only the final block can have n < blockSize.
 	n int
-	// started is whether Next has been called at all.
-	started bool
 	// recovering is true when recovering from corruption.
 	recovering bool
 	// last is whether the current chunk is the last chunk of the record.
@@ -225,8 +223,14 @@ func (r *Reader) nextChunk(wantFirst bool) error {
 
 				logNum := binary.LittleEndian.Uint32(r.buf[r.end+7 : r.end+11])
 				if logNum != r.logNum {
-					// Treat a record from a previous instance of the log as EOF.
-					return io.EOF
+					if wantFirst {
+						// If we're looking for the first chunk of a record, we can treat a
+						// previous instance of the log as EOF.
+						return io.EOF
+					}
+					// Otherwise, treat this chunk as invalid in order to prevent reading
+					// of a partial record.
+					return ErrInvalidChunk
 				}
 
 				chunkType -= (recyclableFullChunkType - 1)
@@ -257,11 +261,11 @@ func (r *Reader) nextChunk(wantFirst bool) error {
 			r.recovering = false
 			return nil
 		}
-		if r.n < blockSize && r.started {
-			if r.end != r.n {
-				// This can happen if the previous instance of the log ended with a partial block
-				// at the same blockNum as the new log but extended beyond the partial block of the
-				// new log.
+		if r.n < blockSize && r.blockNum >= 0 {
+			if !wantFirst || r.end != r.n {
+				// This can happen if the previous instance of the log ended with a
+				// partial block at the same blockNum as the new log but extended
+				// beyond the partial block of the new log.
 				return ErrInvalidChunk
 			}
 			return io.EOF
@@ -288,7 +292,6 @@ func (r *Reader) Next() (io.Reader, error) {
 	if r.err != nil {
 		return nil, r.err
 	}
-	r.started = true
 	return singleReader{r, r.seq}, nil
 }
 
@@ -355,7 +358,7 @@ func (r *Reader) seekRecord(offset int64) error {
 
 	// Clear the state of the internal reader.
 	r.begin, r.end, r.n = 0, 0, 0
-	r.started, r.recovering, r.last = false, false, false
+	r.blockNum, r.recovering, r.last = -1, false, false
 	if r.err = r.nextChunk(false); r.err != nil {
 		return r.err
 	}

--- a/internal/record/record_test.go
+++ b/internal/record/record_test.go
@@ -38,10 +38,7 @@ type recordWriter interface {
 }
 
 func testGeneratorWriter(
-	t *testing.T,
-	reset func(),
-	gen func() (string, bool),
-	newWriter func(io.Writer) recordWriter,
+	t *testing.T, reset func(), gen func() (string, bool), newWriter func(io.Writer) recordWriter,
 ) {
 	buf := new(bytes.Buffer)
 
@@ -873,6 +870,19 @@ func TestSize(t *testing.T) {
 	}
 }
 
+type limitedWriter struct {
+	io.Writer
+	limit int
+}
+
+func (w *limitedWriter) Write(p []byte) (n int, err error) {
+	w.limit--
+	if w.limit < 0 {
+		return len(p), nil
+	}
+	return w.Writer.Write(p)
+}
+
 func TestRecycleLog(t *testing.T) {
 	const min = 16
 	const max = 4096
@@ -897,7 +907,13 @@ func TestRecycleLog(t *testing.T) {
 	// with random data.
 	backing := make([]byte, 1<<20)
 	for i := 1; i <= 100; i++ {
-		w := NewLogWriter(bytes.NewBuffer(backing[:0]), uint64(i))
+		blocks := rnd.Intn(100)
+		limitedBuf := &limitedWriter{
+			Writer: bytes.NewBuffer(backing[:0]),
+			limit:  blocks,
+		}
+
+		w := NewLogWriter(limitedBuf, uint64(i))
 		sizes := make([]int, 10+rnd.Intn(100))
 		for j := range sizes {
 			data := randBlock()
@@ -914,10 +930,18 @@ func TestRecycleLog(t *testing.T) {
 		for j := range sizes {
 			rr, err := r.Next()
 			if err != nil {
+				// If we limited output then an EOF, zeroed, or invalid chunk is expected.
+				if limitedBuf.limit < 0 && (err == io.EOF || err == ErrZeroedChunk || err == ErrInvalidChunk) {
+					break
+				}
 				t.Fatalf("%d/%d: %v", i, j, err)
 			}
 			x, err := ioutil.ReadAll(rr)
 			if err != nil {
+				// If we limited output then an EOF, zeroed, or invalid chunk is expected.
+				if limitedBuf.limit < 0 && (err == io.EOF || err == ErrZeroedChunk || err == ErrInvalidChunk) {
+					break
+				}
 				t.Fatalf("%d/%d: %v", i, j, err)
 			}
 			if sizes[j] != len(x) {
@@ -949,6 +973,38 @@ func TestRecycleLogWithPartialBlock(t *testing.T) {
 	if _, err = r.Next(); err != ErrInvalidChunk && err != io.EOF {
 		t.Fatalf("unexpected error: %v", err)
 	}
+}
+
+func TestRecycleLogWithPartialRecord(t *testing.T) {
+	const recordSize = (blockSize * 3) / 2
+
+	// Write a record that is larger than the log block size.
+	backing1 := make([]byte, 2*blockSize)
+	w := NewLogWriter(bytes.NewBuffer(backing1[:0]), uint64(1))
+	_, err := w.WriteRecord(bytes.Repeat([]byte("a"), recordSize))
+	require.NoError(t, err)
+	require.NoError(t, w.Close())
+
+	// Write another record to a new incarnation of the WAL that is larger than
+	// the block size.
+	backing2 := make([]byte, 2*blockSize)
+	w = NewLogWriter(bytes.NewBuffer(backing2[:0]), uint64(2))
+	_, err = w.WriteRecord(bytes.Repeat([]byte("b"), recordSize))
+	require.NoError(t, err)
+	require.NoError(t, w.Close())
+
+	// Copy the second block from the first WAL to the second block of the second
+	// WAL. This produces a scenario where it appears we crashed after writing
+	// the first block of the second WAL, but before writing the second block.
+	copy(backing2[blockSize:], backing1[blockSize:])
+
+	// Verify that we can't read a partial record from the second WAL.
+	r := NewReader(bytes.NewReader(backing2), uint64(2))
+	rr, err := r.Next()
+	require.NoError(t, err)
+
+	_, err = ioutil.ReadAll(rr)
+	require.Equal(t, err, ErrInvalidChunk)
 }
 
 func BenchmarkRecordWrite(b *testing.B) {

